### ISSUE TITLE
9.6 | Performance degradation

### ISSUE THREAD

**Issue by llaville** [overtrue/phplint#225]
After working on issue #218, and begun to work on major version 10.0, I've found origin of performance degradation we may occurs.

I'll keep more details later ...

**Comment by llaville** [comment]
Here is time now to explains what were detected and what was wrong attributed !

With [Cache invalidation](https://github.com/overtrue/phplint/issues/218) issue @staabm reported us a real problem with cache sub-system,
that was solved with versions

- [9.6.0](https://github.com/overtrue/phplint/releases/tag/9.6.0)
- [9.6.1](https://github.com/overtrue/phplint/releases/tag/9.6.1)
- [9.6.2](https://github.com/overtrue/phplint/releases/tag/9.6.2)

We have introduced ability to remove cache entries by expiration time-based invalidation feature.

Next, @staabm [revert back to version 9.5.6](https://github.com/overtrue/phplint/issues/218#issuecomment-2794309687) because he thought that the performance degradation came from pruning cache entries.

This is HERE what was wrong attributed !

The performance degradation does not come from caching feature, but exists from a long time.

Personally I didn't detect it because I scan project with few files or with a subset of their files.
And I've always a fresh cache directory (removed before each scan).

OK, now I've take time to give explains of situation, there are perharps some people that don't trust me, or don't really understand the problem.

I'll gave now all steps that allows me to found origin of the performance degradation.

I'll thanks Derick Rethans for his great extension that we should all knows, I've named [xdebug](https://github.com/xdebug/xdebug)

And the most important for us, in this situation, the [Profiling](https://xdebug.org/docs/profiler) feature.

I don't use the solutions suggested : KCacheGrind, QCacheGrind or Webgrind

Because I use PhpStorm as IDE, we have natively a Xdebug Profiler tools included. 

Simply go to `Tools` > `Analyse Xdebug Profiler Snapshot` and open the `cachegrind.out.x` file written by Xdebug.

> [!NOTE]
>
> Source code was PHPLint itself : branch `9.6`, retrieved with command `git clone --branch 9.6 https://github.com/overtrue/phplint.git`  
> 
> Install all dependencies on PHP 8.1 platform 


My first attempt was then to run PHPLint 9.6.2 with the Xdebug Profiler as follow : 

```shell
time php -d "xdebug.mode=profile" bin/phplint -j 5 --no-configuration --no-progress --no-cache -v .
```

Even if there are lot of files (5827) scanned, it took lot of time (2 minutes) to do it !

So, as decided, I've analyzed the `cachegrind.out.x` file generated and after finding that

![Image](https://github.com/user-attachments/assets/c9a6dc3d-035f-46ee-bc71-68aea7182d26)

I thought that the time "lost" (64 seconds / 70% elapsed time) is located at `\Overtrue\PHPLint\Linter::checkProcessRunning()`

But what is the reality ?

No time lost on file loop, but each file linting process take too much time.

So, I've decided to check source code of [PHP-Parallel-Lint](https://github.com/php-parallel-lint/PHP-Parallel-Lint) project, because they don't suffer to bad performance.

For example, on checking PHPLint 9.6 source code with following command : 

```shell
time php -d "xdebug.mode=off" /shared/backups/php/parallel-lint.phar -j 5 --no-progress .
```

I got this kind of result:

```text
PHP 8.1.32 | 5 parallel jobs
Checked 5858 files in 29.5 seconds
Syntax error found in 165 files

<... errors found ...>

real    0m33.634s
user    1m27.408s
sys     0m50.190s
```

A very good performance !!!

And I found finally origin of our issue. We invoke the native PHP linter without `-n` option.

So when you have lot of extensions loaded, as me (because I works on project https://github.com/llaville/php-compatinfo-db),  

```
[PHP Modules]
amqp
apcu
ast
bcmath
bz2
calendar
Core
ctype
curl
date
dba
dom
enchant
exif
fileinfo
filter
ftp
gd
gettext
gmp
hash
http
iconv
igbinary
imap
intl
json
ldap
libxml
lzf
mailparse
mbstring
mcrypt
memcache
memcached
mongodb
msgpack
mysqli
mysqlnd
OAuth
oci8
odbc
openssl
opentelemetry
pcntl
pcre
PDO
pdo_dblib
PDO_Firebird
pdo_mysql
PDO_OCI
PDO_ODBC
pdo_pgsql
pdo_sqlite
pgsql
Phar
posix
pspell
raphf
readline
redis
Reflection
session
shmop
SimpleXML
snmp
soap
sockets
sodium
solr
SPL
sqlite3
ssh2
standard
sysvmsg
sysvsem
sysvshm
tidy
tokenizer
uploadprogress
uuid
xdebug
xhprof
xlswriter
xml
xmldiff
xmlreader
xmlrpc
xmlwriter
xsl
yac
yaml
yar
Zend OPcache
zip
zlib

[Zend Modules]
Xdebug
Zend OPcache
```

To verify that it was origin of issue, I've applied following patch on PHPLint 9.6.2 source code 

```
diff --git a/src/Linter.php b/src/Linter.php
index 7347e3c..cf2e798 100644
--- a/src/Linter.php
+++ b/src/Linter.php
@@ -258,6 +258,7 @@ final class Linter
         }

         $command[] = '-l';
+        $command[] = '-n';
         array_push($command, ...$files);

         return (new LintProcess($command))->setFiles($files);
```

And execute scan again (without xdebug profiler, to gain a little time)

```shell
time php -d "xdebug.mode=off" bin/phplint -j 5 --no-configuration --no-progress --no-cache -v .
```



```text
phplint 9.6.x-dev@e2ebdba by overtrue and contributors.

Runtime       : PHP 8.1.32
Configuration : No config file loaded

┌──────────────────┬──────────────────┐
│ Name             │ Value            │
├──────────────────┼──────────────────┤
│ path             │ ["."]            │
├──────────────────┼──────────────────┤
│ configuration    │ ".phplint.yml"   │
│ no-configuration │ true             │
│ exclude          │ []               │
│ extensions       │ ["php"]          │
│ jobs             │ 5                │
│ no-cache         │ true             │
│ cache            │ ".phplint.cache" │
│ cache-dir        │ ".phplint.cache" │
│ cache-ttl        │ 3600             │
│ no-progress      │ true             │
│ progress         │ "printer"        │
│ output           │ null             │
│ format           │ ["console"]      │
│ warning          │ false            │
│ memory-limit     │ "512M"           │
│ ignore-exit-code │ false            │
│ bootstrap        │ ""               │
└──────────────────┴──────────────────┘


Time: 36 secs, Memory: 68.0 MiB, Cache: 0 hit, 5827 misses, Processes: 5827


 [ERROR] 5827 files, 165 errors

<... errors found ...>

real    0m37.658s
user    1m45.434s
sys     1m21.482s
```

A better performance, isn't it ?

Last but not least, if I use our cache subsystem (removing `--no-cache` on phplint invocation), I got pretty good performances

With cold cache 

```
real    0m47.375s
user    1m53.592s
sys     1m30.172s
```

And on next invocation, with hot cache

```
real    0m3.295s
user    0m4.227s
sys     0m2.830s
```

**Comment by llaville** [comment]
The fix was applied on branch `9.6` with commit https://github.com/overtrue/phplint/commit/2000cb155600a3de97cb7d3faa5ef5cae4d212da

/cc @overtrue I've also upgraded the [benchmark workflow](https://github.com/overtrue/phplint/blob/9.6/.github/workflows/benchmark.yml) since the first version I've used on issue #197.

Now here are results of the Benchmark Workflow (New Generation and Run) 
=> https://github.com/overtrue/phplint/actions/runs/14430710759

![Image](https://github.com/user-attachments/assets/c4ba7aa6-0cc9-433e-b159-6958f9897374)

![Image](https://github.com/user-attachments/assets/6f120c03-8fa5-441e-86c3-2c19f11d876e)

For those who don't have time to download each artifact (4 zip files), here are screenshot of each `index.html` you can find on each zip file.

Of course PHP 8.3 and 8.4 and more efficient because we use native multiple file syntax checks at once, available since PHP 8.3 
=> https://php.watch/versions/8.3/cli-lint-multiple-files

How to read the graphs : 

- baseline bar correspond to version 9.6.2
- fixed bar correspond to version 9.6.2 + patch (see commit referenced on top; current branch 9.6)

## PHP 8.1 

![Image](https://github.com/user-attachments/assets/b01ae5d2-1cc1-49c9-9b30-5b9fcf9016fd)

## PHP 8.2 

![Image](https://github.com/user-attachments/assets/755d324b-d6f5-44b8-8fb9-cf44e8e7aef7)

## PHP 8.3

![Image](https://github.com/user-attachments/assets/b84581f6-440a-40f0-90fd-bfd1ffeac4fc)

## PHP 8.4

![Image](https://github.com/user-attachments/assets/2fcbc1e7-a784-4747-add0-66159f1e3ca3)

**Comment by llaville** [comment]
Latest run with PHP 8.5 comparaison included available at https://github.com/overtrue/phplint/actions/runs/19726965882
